### DISCUSSION TITLE
The memory usage is huge on MQTT with cleanSession=false and can not be limited by high water mark (3.12.13)

### DISCUSSION THREAD

**Discussion by absolute8511** [rabbitmq/rabbitmq-server#10627] [Other] [answered]
### Describe the bug

As showed below, the mqtt reader seems using more than 10MB for each connection, and this memory can not be limited by high watermark setting @ansd 

![e0gPbDYooB](https://github.com/rabbitmq/rabbitmq-server/assets/678699/87f2077d-958f-4413-b974-307a9b6e0c2e)

```
Memory
Total memory used: 11.711 gb
Calculation strategy: rss
Memory high watermark setting: 0.7 of available memory, computed to: 22.5486 gb
connection_other: 42.0437 gb (85.27 %)
allocated_unused: 5.4495 gb (11.05 %)
other_proc: 0.8869 gb (1.8 %)
mnesia: 0.2494 gb (0.51 %)
queue_procs: 0.2216 gb (0.45 %)
binary: 0.1292 gb (0.26 %)
plugins: 0.113 gb (0.23 %)
other_system: 0.0682 gb (0.14 %)
other_ets: 0.0536 gb (0.11 %)
metrics: 0.0514 gb (0.1 %)
code: 0.0367 gb (0.07 %)
atom: 0.0015 gb (0.0 %)
msg_index: 0.0004 gb (0.0 %)
quorum_ets: 0.0 gb (0.0 %)
mgmt_db: 0.0 gb (0.0 %)
quorum_queue_dlx_procs: 0.0 gb (0.0 %)
quorum_queue_procs: 0.0 gb (0.0 %)
stream_queue_procs: 0.0 gb (0.0 %)
stream_queue_replica_reader_procs: 0.0 gb (0.0 %)
connection_channels: 0.0 gb (0.0 %)
connection_readers: 0.0 gb (0.0 %)
connection_writers: 0.0 gb (0.0 %)
queue_slave_procs: 0.0 gb (0.0 %)
reserved_unallocated: 0.0 gb (0.0 %)
stream_queue_coordinator_procs: 0.0 gb (0.0 %)
File Descriptors
Total: 13508, limit: 1048479
Sockets: 13506, limit: 943629
Free Disk Space
Low free disk space watermark: 2.0 gb
Free disk space: 105.9739 gb
Totals
Connection count: 13506
Queue count: 76494
Virtual host count: 1
```

### Reproduction steps

1. using the newest rabbitmq 3.12.13
2. using the server config with below, with  5 nodes
```
rabbitmq:
    additionalConfig: |
      cluster_partition_handling = pause_minority
      collect_statistics_interval = 160000
      log.file.rotation.count = 20
      log.file.rotation.date = $D0
      tcp_listen_options.backlog = 512
      tcp_listen_options.keepalive = true
      tcp_listen_options.linger.on = true
      tcp_listen_options.linger.timeout = 0
      tcp_listen_options.nodelay = true
      tcp_listen_options.recbuf = 32768
      tcp_listen_options.sndbuf = 32768
      mqtt.tcp_listen_options.sndbuf  = 4024
      mqtt.tcp_listen_options.recbuf  = 4024
      mqtt.tcp_listen_options.buffer = 4024
      mqtt.mailbox_soft_limit = 10
      vm_memory_high_watermark.relative = 0.7
      classic_queue.default_version = 2
      management_agent.disable_metrics_collector = true
      log.console = true
      log.console.level = warning
    additionalPlugins:
    - rabbitmq_mqtt
    - rabbitmq_web_mqtt
    advancedConfig: |-
      [
        {rabbit, [
            {connection_max, 900000}
        ]}
      ].
    envConfig: |
      RABBITMQ_SERVER_ADDITIONAL_ERL_ARGS="+P 40000000 +zdbbl 1000000 -mnesia dump_log_write_threshold 50000 +S 20:20 +SDcpu 10:10 +SDio 15"
      ERL_MAX_PORTS=5000000
```
3. use bench mark below
```
./emqtt_bench sub --keepalive=600 --num-retry-connect=5 --topic=pod-test2/%i
          --qos=1 -c 20000 -i 4 --host xxx
          --port 1883 --version 4 --username xxx --password xxx --clean false

./emqtt_bench pub --keepalive=600 --num-retry-connect=5 --topic=pod-test2/%i
          --qos=1 --interval_of_msg=120000 --size=64 -c 20000 -i 10 --host xxx
          --port 1883 --version 4 --username xxx --password xxx --clean false;
```
wait for running for a while and everything looks ok, until we start a new bench again below,  we can see the server memory usage will grow quickly
```
./emqtt_bench sub --keepalive=600 --num-retry-connect=5 --topic=pod-test2/%i
          --qos=1 -c 20000 -i 4 --host xxx
          --port 1883 --version 4 --username xxx --password xxx --clean false
```

### Expected behavior

Should not use so much memory and the memory used by mqtt should be limited

**Comment by michaelklishin** [comment]
This topic has been documented for years in [Tuning for a Large Number of Concurrent Connections](https://www.rabbitmq.com/docs/networking#tuning-for-large-number-of-connections-tcp-buffer-size)

**Reply by absolute8511** [reply]
tcp options has already been used in the config above, but it is not work as expected. From the output of `rabbitmq-diagnostics 
  observer` and `rabbitmq-diagnostics  status`, we can see the memory is not used by tcp buffer.
And by the way, in this test the connections is just 60K which is not so much

**Comment by michaelklishin** [comment]
[This blog post](https://www.rabbitmq.com/blog/2023/03/21/native-mqtt) uses very low TCP buffer sizes but for mostly idle connections this should be perfectly fine.

Also, the high VM memory watermark does not limit the number of connections. What it does is documented [in the Alarms guide](https://www.rabbitmq.com/docs/alarms).

There is a [separate mechanism](https://www.rabbitmq.com/docs/vhosts#limits) (more than one, in fact) that allows you to limit the number of connections specifically to protect the node from [applications that leak connections](https://www.rabbitmq.com/docs/connections#monitoring).

**Comment by ansd** [comment]
FWIW, I tried to repro.

I deployed a single node (not 5 node) cluster on a Linux box:
```
make run-broker PLUGINS="rabbitmq_web_mqtt rabbitmq_management rabbitmq_prometheus" RABBITMQ_CONFIG_FILE="rabbitmq.conf" RABBITMQ_SERVER_ADDITIONAL_ERL_ARGS="+S 6 +P 40000000 +zdbbl 1000000 -mnesia dump_log_write_threshold 50000" FULL=1
```
with the following `rabbitmq.conf`:
```
cluster_partition_handling = pause_minority
collect_statistics_interval = 160000
tcp_listen_options.backlog = 512
tcp_listen_options.keepalive = true
tcp_listen_options.linger.on = true
tcp_listen_options.linger.timeout = 0
tcp_listen_options.nodelay = true
tcp_listen_options.recbuf = 32768
tcp_listen_options.sndbuf = 32768
mqtt.tcp_listen_options.sndbuf  = 4024
mqtt.tcp_listen_options.recbuf  = 4024
mqtt.tcp_listen_options.buffer = 4024
mqtt.mailbox_soft_limit = 10
vm_memory_high_watermark.relative = 0.7
classic_queue.default_version = 2
management_agent.disable_metrics_collector = true
connection_max = 900000
```
I started clients as follows:

```
./emqtt_bench sub --keepalive=600 --num-retry-connect=5 --topic=pod-test2/%i --qos=1 -c 20000 -i 4 --port 1883 --version 4 --username guest --password guest --clean false
```

```
./emqtt_bench pub --keepalive=600 --num-retry-connect=5 --topic=pod-test2/%i --qos=1 --interval_of_msg=120000 --size=64 -c 20000 -i 10 --port 1883 --version 4 --username guest --password guest --clean false
```

after 2 mins:
```
./emqtt_bench sub --keepalive=600 --num-retry-connect=5 --topic=pod-test2/%i --qos=1 -c 20000 -i 4 --port 1883 --version 4 --username guest --password guest --clean false
```

Messages flow through the system and memory usage of the broker is kept at only 3.2 GB.
Memory usage per MQTT connection is kept below 66 KB.

**Reply by absolute8511** [reply]
With 5 nodes is there any different?  I can reproduce in the same test, and when I create the sub connections more slowly (<100/s) the memory usage can be controlled.

**Reply by absolute8511** [reply]
And I saw the high memory usage may go down eventually if the sub connections can create success.

**Reply by absolute8511** [reply]
I dumped the memory call stack as below, it seems create queue files is slow 

![MMVuPOrF8k](https://github.com/rabbitmq/rabbitmq-server/assets/678699/79cf8fab-2da3-428a-ab19-d75fbfea20f0)

![image](https://github.com/rabbitmq/rabbitmq-server/assets/678699/5838d9ae-fd77-490f-b0af-839a7239eeb0)

**Reply by ansd** [reply]
@absolute8511 thanks for the follow up, that's good debugging!

I was about to ask whether 1 vs 5 nodes make a difference for you.
Mnesia transactions in a 5 nodes cluster are indeed more expensive compared to a 1 node cluster.
Creating a queue and binding per subscriber results in Mnesia sync transactions across all nodes, which may be slow.

Some follow up questions I have:
1. Do you hit a memory alarm during your test? (You can observe memory alarms in the RabbitMQ log.)
2. In your case the MQTT connections with high memory usage seem to be executing Mnesia transactions. Nevertheless, why is the memory usage so high with 9MB per connection? In your dev env, can you run `erlang:process_info(MqttConnectionPid, [backtrace, binary, current_function, current_stacktrace, dictionary, garbage_collection, garbage_collection_info, heap_size, memory, message_queue_len, messages, min_heap_size, min_bin_vheap_size, message_queue_data, stack_size, status, total_heap_size])` and additionally `sys:get_status(MqttConnectionPid)` and `sys:get_state(MqttConnectionPid)` so that we understand in more detail where the 9MB memory usage come from?
3. Can you please test the same scenario in RabbitMQ 3.13.0 and [Khepri](https://rabbitmq.github.io/khepri/) instead of Mnesia (i.e. feature flag `khepri_db` enabled)? Does Khepri perform better?

**Reply by absolute8511** [reply]
1. it may hit a memory alarm if running long enough and will grow even after alarm and finally got oom.
2. I tried a few times but I can not catch the info, because it quickly got disappeared after I enter the command. I will try more later.
3. I need do some work to get 3.13 working and will try later. However, we may need make sure the old Mnesia can work as expected even it may slower. We can afford the slower mqtt creation as long as the memory will not grow too much.

**Reply by absolute8511** [reply]
[mqtt.details.txt](https://github.com/rabbitmq/rabbitmq-server/files/14434607/mqtt.details.txt)

I dumped all the process_info, hope it can be useful for you  @ansd 

the file is generated by code below, maybe we can enhance the `rabbitmq_top` plugin to get these info more quickly in future.

```
Pids = processes(),
Details = lists:filtermap(fun(Pid) ->
    try
        {true, {Pid, erlang:process_info(Pid, [backtrace, binary, current_function, current_stacktrace, dictionary, garbage_collection, garbage_collection_info, heap_size, memory, message_queue_len, messages, min_heap_size, min_bin_vheap_size, message_queue_data, stack_size, status, total_heap_size])}}
    catch
        _:_ -> false
    end
end, Pids),
file:write_file("process_details.txt", io_lib:format("~p~n", [Details])).
```

**Reply by absolute8511** [reply]
@ansd any other clues do you need to investigate this memory usage issue?  In 5 nodes cluster, it should be reproduced.

**Reply by ansd** [reply]
Thanks for the infos. Unfortunately the text file you attached doesn't show the processes with high memory usage in your very 1st screenshot.

I haven't had time to repro with 5 nodes.

> I need do some work to get 3.13 working and will try later.

My expectation is that Khepri will perform better. Can you confirm? What were you findings with Khepri? How did it compare to Mnesia?

**Reply by absolute8511** [reply]
Any suggestion to dump the memory usage you want to check? Seems the pids returned by processes() did not include the mqtt reader

**Reply by absolute8511** [reply]
I tried the 3.13.0 with Khepri enabled, it showed the same issue as below

![image](https://github.com/rabbitmq/rabbitmq-server/assets/678699/09d4618f-708f-4165-b375-d92908650c17)


![image](https://github.com/rabbitmq/rabbitmq-server/assets/678699/83d89c69-3527-45bb-b954-c3431680277d)


![image](https://github.com/rabbitmq/rabbitmq-server/assets/678699/7dfa7564-7385-48b4-bb45-b9963edd3deb)

**Reply by michaelklishin** [reply]
It did not "show the same result" according to the screeshot: the top contributors are very different.

## Before

In the first screenshot, it is a management plugin cache which suggests that the HTTP API is used heavily to query for something. Reducing the frequency and avoiding queries like `GET /api/queues` which return some 80 metrics for **all** queues will help. We very often see this endpoint used for monitoring, often of a single queue, sometimes of a SINGLE METRIC of a single queue, as if it was free to call and consumed no resources. It will return roughly 80 JSON fields (IIRC) per queue, and cache the results for a period of time to avoid transferring this information from peer nodes.

## After

On the second screenshot, the biggest contributor (183 MB) is the Khepri data store, which is expected: it keeps its entire data set in memory as well as on disk.

**Reply by michaelklishin** [reply]
With every connection in your setup taking 12 MB, it does not really matter if there is single large contributor like Khepri that takes 200 MB.

I must be repeating myself but I don't see why [reducing TCP buffer size to 1K or 2K](https://www.rabbitmq.com/docs/networking#tuning-for-large-number-of-connections-tcp-buffer-size) (from the OS-specific autotuning default that can autotune to 100s KB or even a few MBs) would not work for MQTT the same way it has been for years for other protocols.

**Reply by ansd** [reply]
Do you have `queue_master_locator=min-masters` configured as policy or globally?

**Reply by ansd** [reply]
If yes, try again without such a policy, and please report back whether this helps.

**Reply by absolute8511** [reply]
Since the total memory usage is more than 25GB, maybe the biggest 183MB is not an issue? After all connections finished, the total memory usage is fall below to only 3GB.  For each connection taking 12MB, with 2000/s  creation it will cost 24GB at the peak.

**Reply by absolute8511** [reply]
I found the `queue_master_locator=min-masters` is set by default in rabbitmq cluster-operator, so I test with `queue_master_locator=random` again, it seems much better. Under 2000/s  connection creation it used less than 4GB memory.

**Reply by ansd** [reply, marked as answer]
So, we found the bottleneck:
`queue_master_locator = min-masters` (applicable to classic queues) [reads **all** queue records from the database](https://github.com/rabbitmq/rabbitmq-server/blob/9145b5386d3d7f67ec83b4a93bef35ff11e336b0/deps/rabbit/src/rabbit_queue_location_min_masters.erl#L34-L49), which leads to unnecessary high resource usage.
In contrast, `queue_leader_locator = balanced` (applicable to quorum queues and streams) provides a better implementation since nodes will be selected randomly when there are more than [1000](https://github.com/rabbitmq/rabbitmq-server/blob/9145b5386d3d7f67ec83b4a93bef35ff11e336b0/deps/rabbit/src/rabbit_queue_location.erl#L17) queues.

`queue_master_locator = min-masters` is indeed the [default](https://github.com/rabbitmq/cluster-operator/blob/b6609a7cf1d5df74b3ca0225ae769cab26b3da95/internal/resource/configmap.go#L31) in the cluster-operator.

I suggest you use `queue_master_locator = client-local` so that your MQTT clients can consume locally from the node they are connected to.

> I test with queue_master_locator=random again, it seems much better. Under 2000/s connection creation it used less than 4GB memory

@absolute8511 did you test `queue_master_locator=random` with Khepri or with Mnesia? Did you observe better performance with Khepri?

**Reply by ansd** [reply]
created https://github.com/rabbitmq/rabbitmq-website/pull/1826

**Reply by absolute8511** [reply]
I test both with Khepri and with Mnesia, and it showed not much different for memory usage
